src/conditions.lisp

```diff
@@ -20,29 +22,154 @@
 
 (/debug "loading conditions.lisp!")
 
+;;; Condition
+(eval-when (:compile-toplevel :load-toplevel :execute)
+  (defun %def-condition (name parents slots options)
+    (let*
+        ((cpl (if (and parents) (remove name parents) '(condition)))
+         (report (if (and options (assoc :report options))
+                     (cadr (assoc :report options))))
+         (class-options (if (and options)
+                            (remove :report options :key 'car)))
+         (class `(defclass ,name
+                     ,(progn
+                        (if (null cpl) '(condition) cpl))
+                   ,slots ,class-options))
+         (report-fn)
+         (method))
+      (when report
+        (typecase report
+          (string
+           (setq report-fn `(lambda (condition stream)
+                              (write-string ,report stream)
+                              ,report)))
+          (list
+           (if (not (eql (car report) 'lambda))
+               (error "Must be LAMBDA ~s." report))
+           (setq report-fn `,report))
+          (t (error "What can I compile it: ~s ?" report)))
+        (setq method
+              `(defmethod print-object ((condition ,name) &optional (stream *standard-output*))
+                 (if *print-escape*
+                     (call-next-method)
+                     (funcall #',report-fn condition stream))
+                 nil)))
+      (values name class method))))
+
+(defmacro define-condition (name (&rest parents) (&rest slot-spec) &rest options)
```

Comment:
Very nice!

Is this easily tested within the host implementation as well though?   I foudn really useful to be able to load all modules in the host system.

I did this by prefixing standard macro and functions like

`(defmacro %define-condition ...)`

and defining later an alias for them conditionally

`#+jscl (defmacro define-condition ..)`

That way debugging a lot easier!